{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41ececc1-ee9b-4f8d-a12e-226c269eca13",
   "metadata": {},
   "outputs": [],
   "source": [
    "#check the input, prevent waste of resources.\n",
    "def check_before_task(taxonomy,protein_list):\n",
    "    os.chdir('/data/songshuang/server_find_novel_regulator')\n",
    "    if len(protein_list)>40:\n",
    "        return 'Too many proteins, the job is rejected.'\n",
    "    #check taxonomy\n",
    "    os.system('datasets download genome taxon %s --assembly-source refseq --include gbff --reference  --preview > preview'%(taxonomy))\n",
    "    data=open('preview').read()\n",
    "    tax_count=data.split(',')[1].split(':')[1]\n",
    "    print(tax_count)\n",
    "    os.system('rm preview')\n",
    "    if int(tax_count) < 5 or int(tax_count) >600 :\n",
    "        return 'Species count is %d, the job is rejected.'%int(tax_count)\n",
    "    return 'True'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9f3c0e7-516f-444b-b228-d602a3f02fc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download genome files from NCBI\n",
    "import os\n",
    "def download_gbk(taxonomy):\n",
    "    os.chdir('/data/songshuang/server_find_novel_regulator/')\n",
    "    #make dir\n",
    "    os.system('mkdir /data/songshuang/server_find_novel_regulator/0_gbk/'+taxonomy)\n",
    "    os.system('mkdir /data/songshuang/server_find_novel_regulator/0_protein/'+taxonomy)\n",
    "    os.system('mkdir /data/songshuang/server_find_novel_regulator/0_promoter/'+taxonomy)\n",
    "    #download from NCBI\n",
    "    os.chdir('/data/songshuang/server_find_novel_regulator/0_gbk/'+taxonomy)\n",
    "    os.system('datasets download genome taxon %s --assembly-source refseq --include gbff --reference'%(taxonomy))\n",
    "    os.system('unzip ncbi_dataset.zip')\n",
    "    os.system('rm -rf ncbi_dataset.zip README.md ncbi_dataset/data/dataset_catalog.json ncbi_dataset/data/assembly_data_report.jsonl')\n",
    "    \n",
    "    for folder in os.listdir('ncbi_dataset/data'):\n",
    "        genome_id=folder.split('.')[0].split('_')[1]\n",
    "        os.system('mv ncbi_dataset/data/%s/genomic.gbff ./%s'%(folder,genome_id))\n",
    "        \n",
    "    os.system('rm -rf ncbi_dataset')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c594951-1596-4fef-ad38-2fb70e0f932c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#extract protein & promoter\n",
    "import os\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "Entrez.email = \"shuang_s@zju.edu.cn\"\n",
    "def all(genome_id):\n",
    "    try:\n",
    "        protein_data=''\n",
    "        promoterdata=''\n",
    "        #print('running')\n",
    "        for seq_record in SeqIO.parse(genome_id, \"genbank\"):\n",
    "            for feature in seq_record.features:\n",
    "                if feature.type == 'CDS':\n",
    "                    if 'protein_id' not in feature.qualifiers:\n",
    "                        continue\n",
    "                    #print(feature.type)\n",
    "                    #print(feature.location)\n",
    "                    protein_seq = feature.qualifiers['translation'][0]\n",
    "\n",
    "                        \n",
    "                    if '-' in str(feature.location):\n",
    "                        promoter_seq = seq_record.seq[(feature.location.end):(feature.location.end + 200)].reverse_complement()\n",
    "                    else:\n",
    "                        promoter_seq = seq_record.seq[(feature.location.start - 200):(feature.location.start)]\n",
    "                        \n",
    "                    if len(protein_seq) < 30 or len(promoter_seq) < 60:\n",
    "                        continue\n",
    "                        \n",
    "                    try:\n",
    "                        protein_data += '>%s\\n%s\\n' % (genome_id+'__'+feature.qualifiers['protein_id'][0] ,protein_seq)#.split('_')[0]\n",
    "                        promoterdata += '>%s\\n%s\\n' % (genome_id+'__'+feature.qualifiers['protein_id'][0] ,promoter_seq)#.split('_')[0]\n",
    "\n",
    "                    except:\n",
    "                        protein_data += '>%s\\n%s\\n' % ('unknown' , protein_seq)\n",
    "                        promoterdata += '>%s\\n%s\\n' % ('unknown' , promoter_seq)\n",
    "                        \n",
    "\n",
    "\n",
    "        f=open('%s.faa'%(genome_id),'w+')\n",
    "        f.write(protein_data)\n",
    "        f.close()\n",
    "        \n",
    "        f=open('%s.fa'%(genome_id),'w+')\n",
    "        f.write(promoterdata)\n",
    "        f.close()\n",
    "        print('finished')\n",
    "    except:\n",
    "        raise\n",
    "        print(genome_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3412693a-792f-4754-a178-3ed70fa88350",
   "metadata": {},
   "outputs": [],
   "source": [
    "#download user's protein sequences\n",
    "def get_selected_protein(protein,tax):\n",
    "    from Bio import Entrez\n",
    "    from Bio import SeqIO\n",
    "    Entrez.email = \"shuang_s@zju.edu.cn\"\n",
    "    data=Entrez.efetch(db=\"protein\",id=protein,rettype=\"fasta\")\n",
    "    seq_data=SeqIO.read(data,'fasta')\n",
    "    output='>%s\\n%s'%(seq_data.id,seq_data.seq)\n",
    "    return output\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b9ca69d-99f0-4829-a147-ca462ae6aeba",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "#extract protein & promoter to build database\n",
    "def extract_protein(tax):\n",
    "    os.chdir('/data/songshuang/server_find_novel_regulator/0_gbk/'+tax)\n",
    "    from multiprocessing import Pool\n",
    "    p=Pool(32)\n",
    "    p.map(all,os.listdir('/data/songshuang/server_find_novel_regulator/0_gbk/'+tax))\n",
    "    os.system('mv *.fa /data/songshuang/server_find_novel_regulator/0_promoter/'+tax)\n",
    "    os.system('mv *.faa /data/songshuang/server_find_novel_regulator/0_protein/'+tax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dd2fdbb-81d4-4fd3-a163-77bf3874dd16",
   "metadata": {},
   "outputs": [],
   "source": [
    "#build diamond database\n",
    "def diamond_task(tax):\n",
    "    os.chdir('/data/songshuang/server_find_novel_regulator/0_protein/'+tax)\n",
    "    os.system('cat * > all_protein.faa')\n",
    "    os.system('diamond makedb --in all_protein.faa -p 32 -d all_protein.dmnd')\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d74d7123-3350-47d6-8040-17da0adf1785",
   "metadata": {},
   "outputs": [],
   "source": [
    "#find promoter sequence in database\n",
    "import os\n",
    "def find_promoter(genome_id,protein_id,tax,form):\n",
    "    if form=='protein':\n",
    "        f=open('/data/songshuang/server_find_novel_regulator/0_protein/%s/%s.faa'%(tax,genome_id))\n",
    "    else:\n",
    "        f=open('/data/songshuang/server_find_novel_regulator/0_promoter/%s/%s.fa'%(tax,genome_id))\n",
    "    data=''\n",
    "    write=False\n",
    "    for line in f:\n",
    "        if (line[0]=='>'):\n",
    "            if write==True:\n",
    "#                 if len(data)<=50:\n",
    "#                     print('here!')\n",
    "#                print(data)\n",
    "                return data\n",
    "            if (protein_id in line):\n",
    "                write=True\n",
    "        if write==True:\n",
    "            data+=line\n",
    "    return '>%s\\nAAAAAA\\n'%(protein_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c947135-f3c1-49dd-9015-a5ea9c7f8cf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "def get_selected_protein_diamond(protein_list,tax,ntime):\n",
    "    os.system('mkdir /data/songshuang/server_find_novel_regulator/99_User_project/'+ntime)\n",
    "    os.system('mkdir /data/songshuang/server_find_novel_regulator/99_User_project/%s/0_selected_protein'%ntime)\n",
    "    os.system('mkdir /data/songshuang/server_find_novel_regulator/99_User_project/%s/1_diamond_result'%ntime)\n",
    "    os.system('mkdir /data/songshuang/server_find_novel_regulator/99_User_project/%s/2_promoter'%ntime)\n",
    "    os.system('mkdir /data/songshuang/server_find_novel_regulator/99_User_project/%s/3_meme_result'%ntime)\n",
    "    \n",
    "\n",
    "    \n",
    "    import pandas as pd\n",
    "    #diamond\n",
    "    for protein in protein_list:\n",
    "        os.chdir('/data/songshuang/server_find_novel_regulator/99_User_project/%s/0_selected_protein'%ntime)\n",
    "        f=open(protein,'w+')\n",
    "        f.write(get_selected_protein(protein,tax))\n",
    "        f.close()\n",
    "        os.system('diamond blastp -d /data/songshuang/server_find_novel_regulator/0_protein/%s/all_protein.dmnd -q %s -o ../1_diamond_result/%s -t /dev/shm -b10.0 -k500 -p 16'%(tax,protein,protein))\n",
    "\n",
    "    #analysis diamond result\n",
    "        diamond_result=pd.read_csv('../1_diamond_result/%s'%(protein),sep='\\t|;',engine='python',header=None)\n",
    "        diamond_result=diamond_result.loc[diamond_result.iloc[:,10]<1e-10,:]\n",
    "    #extract promoter\n",
    "        f=open('/data/songshuang/server_find_novel_regulator/99_User_project/%s/2_promoter/%s'%(ntime,protein),'w+')\n",
    "        diamond_protein_list=list(diamond_result.iloc[:,1])\n",
    "        for diamond_protein in diamond_protein_list:\n",
    "            genome=diamond_protein.split('__')[0]\n",
    "            protein_id=diamond_protein.split('__')[1]\n",
    "            f.write(find_promoter(genome,protein_id,tax,'promoter'))\n",
    "        f.close()\n",
    "        \n",
    "    ##cd-hit the promoter to remove highly similar sequences\n",
    "        os.chdir('/data/songshuang/server_find_novel_regulator/99_User_project/%s/2_promoter'%ntime)\n",
    "        os.system('cd-hit -i %s -c 0.6 -o cd-hit_%s -T 6 -n 2'%(protein,protein))\n",
    "        os.system('mv cd-hit_%s %s'%(protein,protein))\n",
    "        os.system('rm cd-hit*')\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "314ab58e-557f-49d9-a8cb-4ddd7a9a97b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MEME workflow of one single gene group.\n",
    "def meme_work(ntime,com):\n",
    "    os.chdir('/data/songshuang/server_find_novel_regulator/99_User_project/%s/2_promoter'%(ntime))\n",
    "    file_name='-'.join(com)\n",
    "    #print(file_name)\n",
    "    data=''\n",
    "    for name in com:\n",
    "        data+=open(name).read()\n",
    "    f=open('/dev/shm/%s/%s'%(ntime,file_name),'w+')\n",
    "    f.write(data)\n",
    "    f.close()\n",
    "    \n",
    "    os.chdir('/dev/shm/%s'%ntime)\n",
    "    os.system('/home/songshuang/miniconda3/envs/meme/bin/meme %s -oc /data/songshuang/server_find_novel_regulator/99_User_project/%s/3_meme_result/meme/%s -dna -minw 12 -maxw 25 -mod zoops -evt 1e-35 -p 3 -nmotifs 1'%(file_name,ntime,file_name))\n",
    "    os.system('/home/songshuang/miniconda3/envs/meme/bin/meme %s -oc /data/songshuang/server_find_novel_regulator/99_User_project/%s/3_meme_result/pal/%s -dna -minw 12 -maxw 25 -mod zoops -evt 1e-20 -pal -p 3 -nmotifs 1'%(file_name,ntime,file_name))\n",
    "    #os.system('/home/songshuang/miniconda3/envs/meme/bin/meme %s -oc /data/songshuang/server_find_novel_regulator/99_User_project/%s/3_meme_result/meme/%s -dna -minw 12 -maxw 25 -mod zoops -evt 1e-100 -p 3 -nmotifs 1'%(file_name,ntime,file_name))\n",
    "    #os.system('/home/songshuang/miniconda3/envs/meme/bin/meme %s -oc /data/songshuang/server_find_novel_regulator/99_User_project/%s/3_meme_result/pal/%s -dna -minw 12 -maxw 25 -mod zoops -evt 1e-80 -pal -p 3 -nmotifs 1'%(file_name,ntime,file_name))\n",
    "    \n",
    "    \n",
    "    if len(os.listdir('/data/songshuang/server_find_novel_regulator/99_User_project/%s/3_meme_result/meme/%s'%(ntime,file_name))) < 5:\n",
    "        os.system('rm -r /data/songshuang/server_find_novel_regulator/99_User_project/%s/3_meme_result/meme/%s'%(ntime,file_name))\n",
    "        print(file_name+' deleted')\n",
    "        \n",
    "    if len(os.listdir('/data/songshuang/server_find_novel_regulator/99_User_project/%s/3_meme_result/pal/%s'%(ntime,file_name))) < 5:\n",
    "        os.system('rm -r /data/songshuang/server_find_novel_regulator/99_User_project/%s/3_meme_result/pal/%s'%(ntime,file_name))\n",
    "        print(file_name+' deleted')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "383f2c74-f81f-40e8-aaf5-756afd85465e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Sampling and combination workflow\n",
    "def combine_meme_work(ntime):\n",
    "    work=os.listdir('/data/songshuang/server_find_novel_regulator/99_User_project/%s/2_promoter'%(ntime))\n",
    "    from itertools import combinations\n",
    "    works=combinations(work,4)\n",
    "    os.chdir('/data/songshuang/server_find_novel_regulator/99_User_project/%s/2_promoter'%(ntime))\n",
    "    os.system('mkdir /data/songshuang/server_find_novel_regulator/99_User_project/%s/3_meme_result'%(ntime))\n",
    "    os.system('mkdir /data/songshuang/server_find_novel_regulator/99_User_project/%s/3_meme_result/meme'%(ntime))\n",
    "    os.system('mkdir /data/songshuang/server_find_novel_regulator/99_User_project/%s/3_meme_result/pal'%(ntime))\n",
    "    os.system('mkdir /dev/shm/%s'%ntime)\n",
    "    from multiprocessing import Pool\n",
    "    from functools import partial\n",
    "    p=Pool(30)\n",
    "    p.map(partial(meme_work, ntime),works)\n",
    "    os.system('rm -r /dev/shm/%s'%ntime)\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39a09a4f-2fff-4770-890e-8d80b2ef0016",
   "metadata": {},
   "outputs": [],
   "source": [
    "#merge similar result and provide a summary\n",
    "def motif_for_R(ntime):\n",
    "    os.chdir('/data/songshuang/server_find_novel_regulator/99_User_project/%s/3_meme_result/meme'%(ntime))\n",
    "    os.system('/home/songshuang/.conda/envs/R/bin/Rscript /home/songshuang/code/DdaA文章/99_server_for_novel/motif.R')\n",
    "    summary_R_result()\n",
    "    os.chdir('/data/songshuang/server_find_novel_regulator/99_User_project/%s/3_meme_result/pal'%(ntime))\n",
    "    os.system('/home/songshuang/.conda/envs/R/bin/Rscript /home/songshuang/code/DdaA文章/99_server_for_novel/motif.R')\n",
    "    summary_R_result()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8297028-ffe5-4f81-b067-d631bcccca92",
   "metadata": {},
   "outputs": [],
   "source": [
    "def summary_R_result():\n",
    "    #summary_result\n",
    "    newf=open('summary.txt','w+')\n",
    "    for file in os.listdir():\n",
    "        if 'Motif' in file:\n",
    "            newf.write(file+'\\n')\n",
    "            f=open(file)\n",
    "            for line in f:\n",
    "                newf.write(line)\n",
    "            f.close()\n",
    "    newf.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
